Fixing wait() issue on destructor and change sleep() to shorter non-blocking one in EngineShoutcast

illuusio opened the pull request:
There is possible deadlock in destructor of EngineShoutcast if network drops and 'wait()' never exit. Change it non-blocking waiting 'while()' system so others can exit while we wait. Change sleep to use QThread::msleep which should be non-blocking and make waiting very short.

illuusio committed:
Changed wait() to non-blocking sleep for waiting that thread exits in 'EngineShoutcast::~EngineShoutcast()'

illuusio committed:
Changed sleep (which actually is also branched from QThread so it doesn't block but waiting for 1 second is in computer time too long) to QThread::msleep. There should be small delay to let network or server to get act back together

illuusio committed:
Decided that max ~10 second wait is too much so lowered it to ~2 seconds. Most time if server is not responding in 2 seconds or network is down for that long we just don't mind about it.

illuusio committed:
Correct intending and coding conventions but only related current issue

daschuer commented on src/engine/sidechain/engineshoutcast.cpp:
```diff
@@ -96,9 +96,26 @@ EngineShoutcast::EngineShoutcast(ConfigObject<ConfigValue>* _config)
+    // Because Thread QThread msec doesn't blocking
+    // Wait 20 * 100 msec which is ~2 seconds for disconnection. If anything ain't happening
+    // it's not going to happen!
+    while (l_iTime < 20) {
```
This burns some CPU cycles. Why we cant use `wait(2000)` here? 

If we assume that the thread does not die reliable once we have set m_pShoutcastEnabled->set(0);
We have also a problem when the user disconnects from shoutcast server. 
What happens if we call later start() again, but we have still locked thread around? 

This issue is probably more annoying than the deadlock at shout down, which is "fixed" here.

illuusio replied:
Actually I think I'll change this wait(2000).

daschuer commented on src/engine/sidechain/engineshoutcast.cpp:
```diff
@@ -401,7 +418,7 @@ bool EngineShoutcast::processConnect() {
         m_iShoutFailures++;
         qDebug() << "Streaming server failed connect. Failures:" << shout_get_error(m_pShout);
-        sleep(1);
+        QThread::msleep(100);
```
I Think we do not need include #include "util/sleep.h" any-more

daschuer replied:
Do you have a clue how may cycles in normal here? 
Please add a comment.

illuusio replied:
Actually these sleeps are just for if there is network/server problem. Wait a small amount time to make network/server settle.

daschuer commented on src/engine/sidechain/engineshoutcast.cpp:
```diff
@@ -412,7 +429,7 @@ bool EngineShoutcast::processConnect() {
             setState(NETWORKSTREAMWORKER_STATE_WAITING);
             qDebug() << "Connection pending. Sleeping...";
-            sleep(1);
+            QThread::msleep(100);
```
The same here, since you reduce the all over waiting time by 10.

illuusio committed:
While was too CPU burner so changed it to more convienient 'wait(4000)' and still not having deadlock

illuusio commented:
Pushed some changes that should now make this actually better. Reading little bit libshout code and noted that aether get connection or end up to misery. Thinking this through made me sure there is no need for these extra delays because trying to open socket will take care of this (if network is slow and if it's fast it doesn't matter anyway).

illuusio committed:
There actually no need for waiting. Removed 'sleep()' because slow network and login to
Shoutcast server will handle random delay. More verbose error handling will give more secure
exit manoveur.

illuusio committed:
In libshout code 'shout_get_connected()' returns error
or tries to reconnect server and that handles sleeping

illuusio committed:
With Icecast2 server longest time seems to be when you have login  and server is getting
things done for you. With test server with no load this tooks around ~1 second.
So we delay only when socket is BUSY to minimize interrupt and signal correct errors for
debug for reasonable bug reports

daschuer commented on src/engine/sidechain/engineshoutcast.cpp:
```diff
@@ -98,7 +98,15 @@ EngineShoutcast::EngineShoutcast(ConfigObject<ConfigValue>* _config)
+    wait(4000);
+
+    // Signal user if thread doesn't die
+    if (isRunning()) {
```
IMHO this should be an 
DEBUG_ASSERT_AND_HANDLE(!isRunning()) {

Since it is a workaround for pending bug, which is also bad if you try to disconnect during a Mixxx run.

daschuer commented on src/engine/sidechain/engineshoutcast.cpp:
```diff
@@ -98,7 +98,15 @@ EngineShoutcast::EngineShoutcast(ConfigObject<ConfigValue>* _config)
+
+    // Signal user if thread doesn't die
+    if (isRunning()) {
+        qDebug() << "EngineShoutcast:~EngineShoutcast(): Thread didn't die. Ignored but add this to bug report if problems rise!";
```
qWarning()

daschuer commented:
Well done, I have just tested it and it works more agile than master. 
I could also not reproduce the lock-out and not the thread dead lock at tread end. 
I added some minor comments, else LGTM.

daschuer commented:
By the way: 
Here the log of the issue where reconnection was not possible: 
```
Debug [EngineShoutcast 1]: shout_queuelen 484846 
Debug [EngineShoutcast 1]: shout_queuelen 487771 
Debug [EngineShoutcast 1]: shout_queuelen 488607 
Debug [EngineShoutcast 1]: shout_queuelen 491115 
Debug [EngineShoutcast 1]: shout_queuelen 491115 
Debug [EngineShoutcast 1]: shout_queuelen 494041 
Debug [EngineShoutcast 1]: EngineShoutcast::processDisconnect() 
Debug [EngineShoutcast 1]: Unloaded libmp3lame  
Debug [EngineShoutcast 1]: EngineShoutcast: updating from preferences 
Debug [EngineShoutcast 1]: Couldn't find shoutcast metadata codec for codec: ""  defaulting to ISO-8859-1. 
Debug [EngineShoutcast 1]: Successfully loaded encoder library  "mp3lame" 
Debug [EngineShoutcast 1]: Loaded libmp3lame version  3.99.5 
Debug [EngineShoutcast 1]: Connection pending. Sleeping... 
Debug [EngineShoutcast 1]: EngineShoutcast::processConnect() socket error. Is socket already in use? 
Debug [EngineShoutcast 1]: Unloaded libmp3lame  
Warning [EngineShoutcast 1]: Streaming error:  "Bitte überprüfen Sie die Internetverbindung und vergewissern sich, dass Ihr Benutzername und Kennwort richtig sind." 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 1 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 2 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 1 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 2 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 1 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 1 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 2 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 1 
Debug [Main]: DlgPrefShoutcast::shoutcastEnabledChanged() 0 
Debug [Main]: Destroying MixxxMainWindow 
```

illuusio commented:
@daschuer Interesting log. Seem that it runs some kind off connection bersek and tries to connect but then disconnects right a way. Have this happen after that?

illuusio committed:
Fixed assetiong in thread handling Shoutcast destrctor

daschuer commented:
All the `DlgPrefShoutcast::shoutcastEnabledChanged` entries where triggered manually. 

No, I did not experienced this again.

daschuer commented:
LGTM. Thank you.

rryan commented:
```
Warning [Main]: DEBUG ASSERT: "isRunning()" in file src/engine/sidechain/engineshoutcast.cpp, line 107
Warning [Main]: EngineShoutcast:~EngineShoutcast(): Thread didn't die.       Ignored but add this to bug report if problems rise!
```
I hit this on every run of Mixxx now...

daschuer commented:
Great! Please produce a backtrack.thanks

illuusio commented:
I'm also pleased because now we have someone who has problems!
